File: backend/handlers/room.go
package handlers

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"net/http"

	"chat-app/room"

	"github.com/gin-gonic/gin"
	"github.com/skip2/go-qrcode"
)

type RoomHandler struct {
	Manager *room.Manager
	BaseURL string
}

func NewRoomHandler(manager *room.Manager, baseURL string) *RoomHandler {
	return &RoomHandler{
		Manager: manager,
		BaseURL: baseURL,
	}
}

type CreateRoomResponse struct {
	RoomID    string `json:"room_id"`
	Token     string `json:"token"`
	ChatURL   string `json:"chat_url"`
	StreamURL string `json:"stream_url"`
	QRUrl     string `json:"qr_url"`
}

func (h *RoomHandler) CreateRoom(c *gin.Context) {
	roomID := generateRoomID()
	h.Manager.CreateRoom(roomID)

	// Create first invite
	invite := h.Manager.CreateInvite(roomID)

	response := CreateRoomResponse{
		RoomID:    roomID,
		Token:     invite.Token,
		ChatURL:   fmt.Sprintf("%s/c/%s", h.BaseURL, invite.Token),
		StreamURL: fmt.Sprintf("%s/s/%s", h.BaseURL, roomID),
		QRUrl:     fmt.Sprintf("/api/qr/%s", invite.Token),
	}

	c.JSON(http.StatusCreated, response)
}

type AddInviteResponse struct {
	Token   string `json:"token"`
	ChatURL string `json:"chat_url"`
	QRUrl   string `json:"qr_url"`
}

// AddInvite creates a new single-use invite for existing room
func (h *RoomHandler) AddInvite(c *gin.Context) {
	roomID := c.Param("roomId")

	r := h.Manager.GetRoom(roomID)
	if r == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "room not found"})
		return
	}

	invite := h.Manager.CreateInvite(roomID)

	response := AddInviteResponse{
		Token:   invite.Token,
		ChatURL: fmt.Sprintf("%s/c/%s", h.BaseURL, invite.Token),
		QRUrl:   fmt.Sprintf("/api/qr/%s", invite.Token),
	}

	c.JSON(http.StatusCreated, response)
}

func (h *RoomHandler) GetQRCode(c *gin.Context) {
	token := c.Param("token")

	invite := h.Manager.GetInvite(token)
	if invite == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "invalid token"})
		return
	}

	chatURL := fmt.Sprintf("%s/c/%s", h.BaseURL, token)

	png, err := qrcode.Encode(chatURL, qrcode.Medium, 256)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "failed to generate QR"})
		return
	}

	c.Data(http.StatusOK, "image/png", png)
}

func generateRoomID() string {
	bytes := make([]byte, 4)
	rand.Read(bytes)
	return hex.EncodeToString(bytes)
}

File: backend-node/handlers/room.js
const QRCode = require('qrcode');
const { addWord, removeWord, getWords } = require('../utils/filter');

class RoomHandler {
  constructor(manager, baseURL) {
    this.manager = manager;
    this.baseURL = baseURL;
  }

  createRoom = async (req, res) => {
    try {
      const roomId = this.manager.generateRoomId();
      this.manager.createRoom(roomId);

      const invite = this.manager.createInvite(roomId);

      res.status(201).json({
        room_id: roomId,
        token: invite.token,
        chat_url: `${this.baseURL}/c/${invite.token}`,
        stream_url: `${this.baseURL}/s/${roomId}`,
        qr_url: `/api/qr/${invite.token}`
      });
    } catch {
      res.status(500).json({ error: 'Failed to create room' });
    }
  };

  addInvite = async (req, res) => {
    try {
      const room = this.manager.getRoom(req.params.roomId);
      if (!room) return res.status(404).json({ error: 'Room not found' });

      const invite = this.manager.createInvite(req.params.roomId);

      res.status(201).json({
        token: invite.token,
        chat_url: `${this.baseURL}/c/${invite.token}`,
        qr_url: `/api/qr/${invite.token}`
      });
    } catch {
      res.status(500).json({ error: 'Failed to create invite' });
    }
  };

  getQRCode = async (req, res) => {
    try {
      const invite = this.manager.getInvite(req.params.token);
      if (!invite) return res.status(404).json({ error: 'Invalid token' });

      const chatURL = `${this.baseURL}/c/${req.params.token}`;
      const pngBuffer = await QRCode.toBuffer(chatURL, { width: 256 });

      res.setHeader('Content-Type', 'image/png');
      res.send(pngBuffer);
    } catch {
      res.status(500).json({ error: 'Failed to generate QR code' });
    }
  };

  // --- Room list ---

  getRooms = (req, res) => {
    res.json({ rooms: this.manager.getRooms() });
  };

  // --- Pin management ---

  getPinnedMessage = (req, res) => {
    const room = this.manager.getRoom(req.params.roomId);
    if (!room) return res.status(404).json({ error: 'Room not found' });
    res.json({ pinnedMessage: room.pinnedMessage });
  };

  pinMessage = (req, res) => {
    const room = this.manager.getRoom(req.params.roomId);
    if (!room) return res.status(404).json({ error: 'Room not found' });

    const { content, name } = req.body;
    if (!content || !content.trim()) {
      return res.status(400).json({ error: 'Content required' });
    }

    const message = {
      content: content.trim(),
      name: (name || 'Admin').trim(),
      timestamp: new Date().toLocaleTimeString('en-US', {
        hour12: false,
        hour: '2-digit',
        minute: '2-digit'
      })
    };

    room.setPinnedMessage(message);
    res.json({ pinnedMessage: message });
  };

  unpinMessage = (req, res) => {
    const room = this.manager.getRoom(req.params.roomId);
    if (!room) return res.status(404).json({ error: 'Room not found' });

    room.clearPinnedMessage();
    res.json({ success: true });
  };

  // --- Censored words management ---

  getWords = (req, res) => {
    res.json({ words: getWords() });
  };

  addWord = (req, res) => {
    const { word } = req.body;
    if (!word || !word.trim()) {
      return res.status(400).json({ error: 'Word required' });
    }
    const added = addWord(word.trim());
    if (!added) return res.status(409).json({ error: 'Word already exists' });
    res.status(201).json({ words: getWords() });
  };

  removeWord = (req, res) => {
    const word = decodeURIComponent(req.params.word);
    const removed = removeWord(word);
    if (!removed) return res.status(404).json({ error: 'Word not found' });
    res.json({ words: getWords() });
  };
}

module.exports = RoomHandler;
